geode-core/src/main/java/org/apache/geode/internal/cache/wan/parallel/ParallelQueueRemovalMessage.java

```diff
@@ -207,8 +207,12 @@ private void destroyKeyFromBucketQueue(AbstractBucketRegionQueue brq, Object key
       }
       // add the key to failedBatchRemovalMessageQueue.
       // This is to handle the last scenario in #49196
-      brq.addToFailedBatchRemovalMessageKeys(key);
-
+      // But if GII is already completed and FailedBatchRemovalMessageKeys
+      // are already cleared then no keys should be added to it as they will
+      // never be cleared and increase the memory footprint.
+      if (!brq.isFailedBatchRemovalMessageKeysClearedFlag()) {
```

Comment:
Will this logic work with auto-reconnect?  If a member is on the losing side of a partition it will do GII when it rejoins the cluster, right?

Reply:
@metatype let me go through the auto-reconnect code to figure out how the  mechanism works. BRB

Reply:
@metatype 
Going through the documentation about auto-reconnect, when a member disconnects from a distributed system, it shuts down and restarts in a "connecting state". 

When it restarts, all the BucketRegionQueues are created again hence setting the flag to false, does the GII and then cleans out any failed batch removal messages that could have arrived during the GII phase and then sets the  failedBatchRemovalMessageKeysClearedFlag to true.